### PR TITLE
Aggregate Continuation Tokens

### PR DESCRIPTION
**Pull request by bchong95**
# Aggregate Continuation Tokens

## Description

This change provides continuation token support for aggregate queries. Currently aggregate queries are not preempt-able, since we need to drain all the partial aggregate and emit the final result to the user. We also can't emit an partial aggregate, since the user is expecting a single result and does not want to aggregate the partial aggregates. The solution is to return empty pages with the partial aggregate in the continuation token. Since aggregates are done on a single grouping the continuation token is simply:

```JSON
{
   "SingleGroupingPartialAggregate": <SingleGroupingPartialAggregate>,
   "SourceContinuationToken": <SourceContinuationToken>
}
```

Like all continuation token changes we need to be able to serialize the query operator state and resume from it.

A natural follow up item is to support GROUP BY continuation tokens, since it's just multiple groupings.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs [hunk 1]
```diff
@@ -40,6 +45,29 @@ public CosmosElement GetResult()
...
+        }
+
+        public static AverageAggregator Create(string continuationToken)
+        {
+            AverageInfo averageInfo;
+            if (continuationToken != null)
+            {
+                if (!AverageInfo.TryParse(continuationToken, out averageInfo))
+                {
+                    throw new ArgumentException($"Invalid continuation token: {continuationToken}");
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs]: >ArgumentException [](start = 30, length = 17)

CosmosException w/ BadRequest

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/CountAggregator.cs [hunk 2]
```diff
@@ -47,5 +58,30 @@ public CosmosElement GetResult()
         {
             return CosmosNumber64.Create(this.globalCount);
         }
+
+        public string GetContinuationToken()
+        {
+            return this.globalCount.ToString();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/CountAggregator.cs]: >return this.globalCount.ToString(); [](start = 12, length = 35)

specify the culture

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/MinMaxAggregator.cs [hunk 3]
```diff
@@ -136,6 +133,72 @@ public CosmosElement GetResult()
...
+                }
+                else if (continuationToken == MinMaxAggregator.UndefinedContinuationToken)
+                {
+                    globalMinMax = MinMaxAggregator.Undefined;
+                }
+                else
+                {
+                    if (!CosmosElement.TryParse(continuationToken, out globalMinMax))
+                    {
+                        throw new ArgumentException($"Malformed continuation token: {continuationToken}");
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/MinMaxAggregator.cs]: >ArgumentException [](start = 34, length = 17)

same

**Comment by sboshra** [comment]
>     internal abstract class SingleGroupAggregator

CompositeAggregator

---
Refers to: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SingleGroupAggregator.cs:16 in ab86e22. [](commit_id = ab86e22f14e1e1730f3f45c21d3b5728007c17df, deletion_comment = False)

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SingleGroupAggregator.cs [hunk 4]
```diff
@@ -123,14 +132,54 @@ public override CosmosElement GetResult()
...
+            }
+
+            public static SelectListAggregateValues Create(
+                CosmosQueryClient cosmosQueryClient,
+                IReadOnlyDictionary<string, AggregateOperator?> aggregateAliasToAggregateType,
+                string continuationToken)
             {
+                if (!CosmosElement.TryParse(continuationToken, out CosmosElement parsedContinuationToken))
+                {
+                    throw cosmosQueryClient.CreateBadRequestException(
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SingleGroupAggregator.cs]: >CreateBadRequestException [](start = 44, length = 25)

use this exception everywhere

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/PipelinedDocumentQueryExecutionContext.cs [hunk 5]
```diff
@@ -134,12 +131,14 @@ public override bool TryGetContinuationToken(out string state)
...
         /// <param name="initParams">The initial parameters</param>
         /// <param name="requestContinuationToken">The request continuation.</param>
+        /// <param name="isComputeQuery">Whether or not the query is suppposed to run on the compute gateway.</param>
         /// <param name="cancellationToken">The cancellation token.</param>
         /// <returns>A task to await on, which in turn returns a CosmosPipelinedItemQueryExecutionContext.</returns>
         public static async Task<CosmosQueryExecutionContext> CreateAsync(
             CosmosQueryContext queryContext,
             CosmosCrossPartitionQueryExecutionContext.CrossPartitionInitParams initParams,
             string requestContinuationToken,
+            bool isComputeQuery,
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/PipelinedDocumentQueryExecutionContext.cs]: >bool isComputeQuery, [](start = 12, length = 20)

consider adding an enum for ExecutionEnvironment (Client, Compute...etc.)

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.cs [hunk 6]
```diff
@@ -64,91 +62,46 @@ internal sealed class AggregateDocumentQueryExecutionComponent : DocumentQueryEx
...
         /// </summary>
+        /// <param name="queryClient">The query client.</param>
         /// <param name="aggregates">The aggregates.</param>
         /// <param name="aliasToAggregateType">The alias to aggregate type.</param>
         /// <param name="hasSelectValue">Whether or not the query has the 'VALUE' keyword.</param>
+        /// <param name="isComputeGateway">Whether or not this is meant to run in the compute gateway.</param>
         /// <param name="requestContinuation">The continuation token to resume from.</param>
         /// <param name="createSourceCallback">The callback to create the source component that supplies the local aggregates.</param>
         /// <returns>The AggregateDocumentQueryExecutionComponent.</returns>
         public static async Task<AggregateDocumentQueryExecutionComponent> CreateAsync(
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.cs]: >AggregateDocumentQueryExecutionComponent [](start = 33, length = 40)

return DocumentQueryExecutionComponentBase

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.SDK.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.CosmosElements;
+
+    internal abstract partial class AggregateDocumentQueryExecutionComponent : DocumentQueryExecutionComponentBase
+    {
+        private sealed class SdkAggregateDocumentQueryExecutionComponent : AggregateDocumentQueryExecutionComponent
+        {
+            private SdkAggregateDocumentQueryExecutionComponent(
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.SDK.cs]: >Sdk [](start = 20, length = 3)

Sdk -> Client

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.CosmosElements;
+
+    internal abstract partial class AggregateDocumentQueryExecutionComponent : DocumentQueryExecutionComponentBase
+    {
+        private static readonly List<CosmosElement> EmptyResults = new List<CosmosElement>();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: >List [](start = 32, length = 4)

please use Enumerable.Empty<TResult> instead

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs]: This code will get removed once we move to a TryCreate pattern.

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/CountAggregator.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/CountAggregator.cs]: done.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs (hunk 8 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: AsReadOnly() get's around this while also preserving the ordering.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+                disallowContinuationTokenMessage: null,
+                continuationToken: updatedContinuationToken,
+                diagnostics: sourceResponse.Diagnostics);
+
+            return response;
+        }
+
+        public override bool TryGetContinuationToken(out string state)
+        {
+            if (!this.IsDone)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: Having all the if else makes it really difficult to read and understand. Can we simplify it by doing this?

```
if (!this.IsDone && this.Source.TryGetContinuationToken(out string sourceState))
{
	AggregateContinuationToken aggregateContinuationToken = AggregateContinuationToken.Create(
		this.singleGroupAggregator.GetContinuationToken(),
		sourceState);
	state = aggregateContinuationToken.ToString();
	return true;
}

state = null;
return false;
```

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        private struct AggregateContinuationToken
+        {
+            private const string SingleGroupAggregatorContinuationTokenName = "SingleGroupAggregatorContinuationToken";
+            private const string SourceContinuationTokenName = "SourceContinuationToken";
+
+            private readonly CosmosObject rawCosmosObject;
+
+            private AggregateContinuationToken(CosmosObject rawCosmosObject)
+            {
+                this.rawCosmosObject = rawCosmosObject;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: null check?

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            public override string ToString()
+            {
+                return this.rawCosmosObject.ToString();
+            }
+
+            public string SingleGroupAggregatorContinuationToken
+            {
+                get
+                {
+                    return (this.rawCosmosObject[AggregateContinuationToken.SingleGroupAggregatorContinuationTokenName] as CosmosString).Value;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: What happens if this is null? Is it possible for it to be a number?

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/MinMaxAggregator.cs [hunk 12]
```diff
@@ -136,6 +133,72 @@ public CosmosElement GetResult()
             return result;
         }
 
+        public string GetContinuationToken()
+        {
+            string continuationToken;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/MinMaxAggregator.cs]: Why not remove all the if else and just return the value?
```
if (this.globalMinMax == ItemComparer.MinValue)
{
	return  MinMaxAggregator.MinValueContinuationToken;
}

if (this.globalMinMax == ItemComparer.MaxValue)
{
	return MinMaxAggregator.MaxValueContinuationToken;
}

if (this.globalMinMax == Undefined)
{
	return MinMaxAggregator.UndefinedContinuationToken;
}

return this.globalMinMax.ToString();
```

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs]: I believe right now all the other continuation token exceptions are using the queryClient.CreateBadRequest. Throwing different exceptions for the same bad continuation token is a horrible user experience. All the continuation exceptions should be throwing the same exception so users can handle it.

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/CountAggregator.cs [hunk 13]
```diff
@@ -47,5 +58,30 @@ public CosmosElement GetResult()
...
+        }
+
+        public static CountAggregator Create(string continuationToken)
+        {
+            long partialCount;
+            if (continuationToken != null)
+            {
+                if (!long.TryParse(continuationToken, out partialCount))
+                {
+                    throw new CosmosException(
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/CountAggregator.cs]: All of this is going to be moot after TryCreatePipeline.

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/AverageAggregator.cs]: Yes that will be the case once we move to a TryCreate pattern.

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SumAggregator.cs [hunk 14]
```diff
@@ -60,5 +65,25 @@ public CosmosElement GetResult()
...
+        {
+            return this.globalSum.ToString("G17");
+        }
+
+        public static SumAggregator Create(string continuationToken)
+        {
+            double partialSum;
+            if (continuationToken != null)
+            {
+                partialSum = double.Parse(continuationToken);
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SumAggregator.cs]: If they move the continuation token to a differently hosted process than there isn't much we can do. They could also move to a process that parses double differently.

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs [hunk 15]
```diff
@@ -130,6 +124,58 @@ public static CosmosElement Parse(string json)
 
             return item;
         }
+
+        public static CosmosElement Parse(string json)
+        {
+            if (!CosmosElement.TryParse(json, out CosmosElement cosmosElement))
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs]: The exception has the full JSON payload in the message.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs (hunk 10 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: good catch! :)

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/MinMaxAggregator.cs (hunk 12 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/MinMaxAggregator.cs]: It's a different value. ItemComparer.MinValue != MinMaxAggregator.MinValueContinuationToken.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs (hunk 11 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: It's an invariant of the class (enforced by TryParse). I will make the invariant stronger and move it to the constructor.

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs [hunk 16]
```diff
@@ -45,12 +45,6 @@ public static CosmosElement CreateFromBuffer(ReadOnlyMemory<byte> buffer)
             return CosmosElement.Dispatch(jsonNavigator, jsonNavigatorNode);
         }
 
-        public static CosmosElement Parse(string json)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs]: Am not yet sure if Gremlin is using them. If so it breaks.

File: Microsoft.Azure.Cosmos/src/Diagnostics/QueryAggregateDiagnostics.cs [hunk 17]
```diff
@@ -26,7 +23,7 @@ internal sealed class QueryAggregateDiagnostics : CosmosDiagnostics
             this.Pages = pages;
         }
 
-        private IReadOnlyCollection<QueryPageDiagnostics> Pages { get; }
+        public IReadOnlyCollection<QueryPageDiagnostics> Pages { get; }
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Diagnostics/QueryAggregateDiagnostics.cs]: It's preference. I view public members in an internal class as internal.

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonTextWriter.cs [hunk 18]
```diff
@@ -398,20 +398,11 @@ private void PrefixMemberSeparator()
...
-                    case '\t':
-                        return true;
-                    default:
-                        return value < ' ';
-                }
+                const char DoubleQuote = '"';
+                const char ReverseSolidus = '\\';
+                const char Space = ' ';
+
+                return (value == DoubleQuote) || (value == ReverseSolidus) || (value < Space);
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonTextWriter.cs]: Let me revert this file.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs (hunk 9 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/AggregateDocumentQueryExecutionComponent.Compute.cs]: You are correct I misread the one bool. Doing the return immediately makes it easier to read and avoid the nested if else logic.  

```
if(this.IsDone)
{
	state = null;
	return true;
}

if (this.Source.TryGetContinuationToken(out string sourceState))
{
	AggregateContinuationToken aggregateContinuationToken = AggregateContinuationToken.Create(
		this.singleGroupAggregator.GetContinuationToken(),
		sourceState);
	state = aggregateContinuationToken.ToString();
	return true;
}

state = null;
return false;
```

File: Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SumAggregator.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/Core/Aggregation/SumAggregator.cs]: I thought the continuation token stored everything as culture invariant so it won't break for the scenario where one machine might be configured to French and the other to English. Are continuation token not support across different cultures?

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs [hunk 19]
```diff
@@ -62,7 +62,8 @@ internal class QueryIterator : FeedIterator
                 MaxItemCount = queryRequestOptions.MaxItemCount,
                 PartitionKey = queryRequestOptions.PartitionKey,
                 Properties = queryRequestOptions.Properties,
-                PartitionedQueryExecutionInfo = partitionedQueryExecutionInfo
+                PartitionedQueryExecutionInfo = partitionedQueryExecutionInfo,
+                ExecutionEnvironment = queryRequestOptions.ExecutionEnvironment.GetValueOrDefault(Core.ExecutionContext.ExecutionEnvironment.Client),
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs]: Re-Serialziation on every invocation.

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 20]
```diff
@@ -141,6 +142,8 @@ public class QueryRequestOptions : RequestOptions
 
         internal CosmosSerializationFormatOptions CosmosSerializationFormatOptions { get; set; }
 
+        internal ExecutionEnvironment? ExecutionEnvironment { get; set; }
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs]: I use get default

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CrossPartitionQueryTests.cs [hunk 21]
```diff
@@ -631,9 +631,12 @@ private CosmosClient CreateNewCosmosClient(ConnectionMode connectionMode)
             string state = null;
             do
             {
+                QueryRequestOptions computeRequestOptions = queryRequestOptions.Clone();
+                computeRequestOptions.ExecutionEnvironment = Cosmos.Query.Core.ExecutionContext.ExecutionEnvironment.Compute;
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CrossPartitionQueryTests.cs]: will do in separate PR once code is merged.

**Pull request merged**